src/mono/mono/metadata/class.c

```diff
@@ -1985,39 +1985,58 @@ mono_class_interface_offset_with_variance (MonoClass *klass, MonoClass *itf, gbo
 
 		return m_class_get_interface_offsets_packed (klass) [found];
 	} else if (has_variance) {
-		MonoClass **vst;
+		MonoVarianceSearchTableEntry *vst;
 		int vst_count;
-		MonoClass *current = klass;
+		mono_class_get_variance_search_table (klass, &vst, &vst_count);
+		int depth = vst_count ? vst[0].depth : 0, j;
+		i = 0;
 
-		// Perform two passes per class, then check the base class
-		while (current) {
-			mono_class_get_variance_search_table (current, &vst, &vst_count);
+		while (depth) {
+			// g_print ("depth==%d, i==%d, count==%d\n", depth, i, vst_count);
 
 			// Exact match pass: Is there an exact match at this level of the type hierarchy?
 			// If so, we can use the interface_offset we computed earlier, since we're walking from most derived to least.
-			for (i = 0; i < vst_count; i++) {
-				if (itf != vst [i])
+			for (j = i; j < vst_count; j++) {
+				if (vst [j].depth != depth)
+					break;
+
+				/*
+				char * cname = mono_type_get_full_name (vst [j].klass);
+				g_print ("  #%d: depth==%d %s\n", j, vst [j].depth, cname);
+				g_free (cname);
+				*/
+				if (itf != vst [j].klass)
 					continue;
 
 				*non_exact_match = FALSE;
+				g_assert (vst [j].offset == exact_match);
 				return exact_match;
 			}
 
 			// Inexact match (variance) pass:
 			// Is any interface at this level of the type hierarchy variantly compatible with the desired interface?
 			// If so, select the first compatible one we find.
-			for (i = 0; i < vst_count; i++) {
-				if (!mono_class_is_variant_compatible (itf, vst [i], FALSE))
+			for (j = i; j < vst_count; j++) {
+				if (vst [j].depth < depth) {
+					i = j;
+					break;
+				}
```

Comment:
Feels like we need a couple comments typing this condition to `build_variance_search_table_inner`.  In particular, how this sets up the "exact match" loop for the next depth to start at the right "`j`".

nit: `i` should have a better name.  it's the "starting index for a run of interfaces from the `klass->idepth - depth`th ancestor of `klass`" (right?).  That should be something other than `i`.

---

I'm not sure I buy `vst[j].depth < detph` as a good "the run of interfaces for the current `depth` has ended here.  Won't this also get into some ancestor's interfaces?  `depth` decreases as `j` goes up, right?

Reply:
I wonder if it would be more straghtforward if we allocated `m_class_get_interface_offsets_count(klass) + placeholders` where `placeholders = m_class_get_idepth (klass)`.  and then we insert `NULL` entries whenever a run for a given ancestor depth is done.

And then all these loops can go over non-null runs of vst entries.

Although I guess that would complicate figuring out when the table is entirely empty.